File: sources/ruby_ship.cpp
#include "ruby_ship.h"

RubyBridge& RubyShip::bridge = RubyBridge::initialize();

RubyShip::RubyShip(const char* class_name): name(class_name)
{
	string cap_name = name;
	cap_name[0] = toupper(cap_name[0]);
	rb_value rb_class = bridge.define_class(name.c_str());
	bridge.exec_file(("ships/"+name+".rb").c_str());
	bridge.exec(("@"+name+" = "+cap_name+".new").c_str());
}

RubyShip::~RubyShip(){}

const char* RubyShip::getName(){ return name.c_str(); }

const string RubyShip::getRubyObj(){ return "@"+name; }

TurnData& RubyShip::getTurnData()
{
	td.energy = stod(bridge.exec(getRubyObj()+".energy"));
	td.x = stoi(bridge.exec(getRubyObj()+".x"));
	td.y = stoi(bridge.exec(getRubyObj()+".y"));
	td.move = stoi(bridge.exec(getRubyObj()+".move"));
	td.turn = stod(bridge.exec(getRubyObj()+".turn"));
	td.max_speed = stoi(bridge.exec(getRubyObj()+".max_speed"));
	td.max_turn_speed = stod(bridge.exec(getRubyObj()+".max_turn_speed"));
	td.radar_turn = stod(bridge.exec(getRubyObj()+".radar_turn"));
	td.radar_max_turn_speed = stod(bridge.exec(getRubyObj()+".radar_max_turn_speed"));
	td.gun_turn = stod(bridge.exec(getRubyObj()+".gun_turn"));
	td.gun_max_turn_speed = stod(bridge.exec(getRubyObj()+".gun_max_turn_speed"));
	
	string s = bridge.exec(getRubyObj()+".fire_power");
	if(s == "nil"){ td.fire_power = 0; td.fire = false; }
	else { td.fire_power = stod(s); td.fire = true; }
	return td;
}

void RubyShip::updateTurnData()
{
	char tmp[256]={};
	sprintf(tmp, ".energy = %g", td.energy);
	bridge.exec("@"+name+tmp);
	sprintf(tmp, ".x = %i", td.x);
	bridge.exec("@"+name+tmp);
	sprintf(tmp, ".y = %i", td.y);
	bridge.exec("@"+name+tmp);
	sprintf(tmp, ".move = %i", td.move);
	bridge.exec("@"+name+tmp);
	sprintf(tmp, ".turn = %g", td.turn);
	bridge.exec("@"+name+tmp);
	sprintf(tmp, ".max_speed = %i", td.max_speed);
	bridge.exec("@"+name+tmp);
	sprintf(tmp, ".max_turn_speed = %g", td.max_turn_speed);
	bridge.exec("@"+name+tmp);
	sprintf(tmp, ".radar_turn = %g", td.radar_turn);
	bridge.exec("@"+name+tmp);
	sprintf(tmp, ".radar_max_turn_speed = %g", td.radar_max_turn_speed);
	bridge.exec("@"+name+tmp);
	sprintf(tmp, ".gun_turn = %g", td.gun_turn);
	bridge.exec("@"+name+tmp);
	sprintf(tmp, ".gun_max_turn_speed = %g", td.gun_max_turn_speed);
	bridge.exec("@"+name+tmp);
}

bool RubyShip::run()
{
	bridge.exec("@"+name+".run");
	return bridge.last_exec();	
}

bool RubyShip::onScannedShip(ScannedShipEvent e)
{
	char tmp[256] = {};
	sprintf(tmp, "event = { bearing: %g, distance: %i, heading: %g, energy: %g, name: \'%s\', speed: %i }\n", e.bearing, e.distance, e.heading, e.energy, e.name, e.speed);
	bridge.exec(string(tmp)+"@"+name+".onScannedShip event");
	return bridge.last_exec();
}

bool RubyShip::onHitByBullet(HitByBulletEvent e)
{
	char tmp[256] = {};
	sprintf(tmp, "event = { bullet: @bullets[%i] }\n", e.bullet_id);
	bridge.exec(string(tmp)+"@"+name+".onHitByBullet event");
	return bridge.last_exec();
}

bool RubyShip::onBulletHit(BulletHitEvent e)
{
	char tmp[256] = {};
	sprintf(tmp, "event = { bullet: @bullets[%i], target_name: '%s', target_energy: %g }\n", e.bullet_id, e.name, e.energy);
	bridge.exec(string(tmp)+"@"+name+".onBulletHit event");
	return bridge.last_exec();
}

bool RubyShip::onBulletHitBullet(BulletHitBulletEvent e)
{
	char tmp[256] = {};
	sprintf(tmp, "event = { bullet: @bullets[%i], bulletHit: @bullets[%i]}\n", e.bullet_id, e.bulletHit_id);
	bridge.exec(string(tmp)+"@"+name+".onBulletHitBullet event");
	return bridge.last_exec();
}

bool RubyShip::onShipHit(ShipHitEvent e)
{
	char tmp[256] = {};
	sprintf(tmp, "event = { bearing: %g, distance: %i, heading: %g, energy: %g, name: '%s' }\n", e.bearing, e.distance, e.heading, e.energy, e.name);
	bridge.exec(string(tmp)+"@"+name+".onShipHit event");	
	return bridge.last_exec();
}

File: ruby_classes/ruby_ship.rb
class RubyShip
	def initialize
		@energy = 100
		@move = 0
		@turn = 0
		@max_speed = 8
		@max_turn_speed = 4
		@radar_turn = 0
		@radar_max_turn_speed = 45
		@gun_turn = 0
		@gun_max_turn_speed = 15
		@lastBulletShot = nil
		@fire_power = nil
		@fired = false
		@y = 0
		@x = 0
	end

	def x
		return @x
	end

	def x= value
		@x = value
	end

	def y
		return @y
	end

	def y= value
		@y = value
	end

	def fire_power
		return @fire_power
	end
	
	def fire_power= value
		return @fire_power
	end
	
	def fired
		return @fired
	end
	
	def fired= value
		return @fired = value
	end

	def energy
		return @energy
	end
	
	def energy= value
		@energy = value
	end
	
	def move
		return @move
	end
	
	def move= value
		return @move = value
	end
	
	def turn
		return @turn
	end
	
	def turn= value
		return @turn = value
	end
	
	def max_speed
		return @max_speed
	end
	
	def max_speed= value
		return @max_speed = value
	end
	
	def max_turn_speed
		return @max_turn_speed
	end
	
	def max_turn_speed= value
		return @max_turn_speed = value
	end
	
	def radar_turn
		return @radar_turn
	end
	
	def radar_turn= value
		return @radar_turn = value
	end
	
	def radar_max_turn_speed
		return @radar_max_turn_speed
	end
	
	def radar_max_turn_speed= value
		return @radar_max_turn_speed = value
	end

	def gun_turn
		return @gun_turn
	end
	
	def gun_turn= value
		return @gun_turn = value
	end
	
	def gun_max_turn_speed
		return @gun_max_turn_speed
	end
	
	def gun_max_turn_speed= value
		return @gun_max_turn_speed = value
	end

	def onScannedShip scannedShipEvent
	end

	def onHitByBullet hitByBulletEvent
	end

	def onBulletHitBullet bulletHitBulletEvent
	end

	def onBulletHit bulletHitEvent
	end
	
	def onShipHit shipHitEvent
	end

	def run
	end

	def setAhead mv
		@move = mv
	end

	def setBack mv
		setAhead -mv
	end

	def setTurnLeft angle
		@turn = angle
	end

	def setTurnRight angle
		setTurnLeft -angle
	end

	def setTurnRadarLeft angle
		@radar_turn = angle
	end

	def setTurnRadarRight angle
		setTurnRadarLeft -angle
	end
	
end
